px.py

```diff
@@ -715,41 +744,111 @@ def do_CONNECT(self):
+                                if out not in outs:
+                                    outs.append(out)
+                                max_idle = time.time() + idle
+                            else:
+                                # No data means connection closed by remote host
+                                dprint("Connection closed by %s" % source)
+                                # Because tunnel is closed on one end there is no need to read from both ends
+                                rlist.clear()
+                                # Do not write anymore to the closed end
+                                if i in wlist:
+                                    wlist.remove(i)
+                                if i in outs:
+                                    outs.remove(i)
+                    if outs:
+                        for o in outs:
+                            if o is self.proxy_socket:
+                                wdata = sdata
+                            else:
+                                wdata = cdata
+                            data = wdata[0]
+                            # socket.send() may sending only a part of the data (as documentation says).
+                            # To ensure sending all data
+                            bsnt = o.send(data)
+                            if bsnt > 0:
+                                if bsnt < len(data):
+                                    # Not all data was sent; store data not sent and ensure select() get's it
+                                    # when the socket can be written again
+                                    wdata[0] = data[bsnt:]
+                                    if o not in wlist:
+                                        wlist.append(o)
+                                else:
+                                    wdata.pop(0)
+                                    if not data and o in wlist:
+                                        wlist.remove(o)
+                                cs += bsnt
+                            else:
+                                dprint("No data sent")
+                        max_idle = time.time() + idle
+                    if max_idle < time.time():
+                        # No data in timeout seconds
+                        dprint("Proxy connection timeout")
+                        break
+
+        # After serving the proxy tunnel it could not be used for samething else.
+        # A proxy doesn't really know, when a proxy tunnnel isn't needed any more (there is no content length for data).
+        # So servings will be ended either after timeout seconds without data transfer or
+        # when at least one side closes the connection.
+        # Close both proxy and client connection if still open.
+        if self.proxy_socket is not None:
+            dprint("Cleanup proxy connection")
+            self.proxy_socket.close()
+            self.proxy_socket = None
+        self.close_connection = True;
```

Comment:
This defeats the whole point of socket reuse - proxy does stay open and reuses connections until you say Proxy-Connection: close. We shouldn't close it or px will slow down since every request will require a new connection.

Reply:
The connection can't be reused because of the http proxy protocol: Client opens a connection to px and sends a CONNECT request. Px opens a connection to the target (DIRECT) or a proxy.  For a proxy it forwards the CONNECT request and handles auth if necessary. Then it uses the client connection and the open (proxy) socket to establish a tunnel between client and target/proxy with a read-write loop. From now on it only forwards all data from client to target/proxy and vice versa. It has no clue what data is transfered and how much. It must not interpret this data and so it can never get another request/response header neither from client nor from target/proxy. It doesn't even know when the tunnel isn't be used anymore other then it detects that either client or target/proxy closes it's end of the connection. That's the reason for the idle timeout: If no data is transfered for this amount of time the tunnel is considered unused and px closes it. So when the read-write loop is left neither the client connection nor the proxy socket can't be used anymore and have to be closed. But that's not a performance problem: Once established the tunnel itself will be used for multiple requests by the client until it no longer needs it and closes the connection or the idle timout is reached..

Reply:
This is true for CONNECT but Px also has to handle plain HTTP in GET, PUT, etc. All those connections can be reused.

Reply:
The code is only in do_CONNECT(). For GET, PUT etc. do_GET() will be called which reuses proxy_socket.